Add Defaulting and Validating Webhooks

wallrj opened the pull request:
Part of: #39

 * Added defaulting and validating webhooks
 * Deploy and configure these in the E2E tests
 * Deploy cert-manager in the E2E tests.

wallrj commented on Makefile:
```diff
@@ -17,9 +17,9 @@ export PATH := $(GOBIN):$(PATH)
+CLEANUP ?= "false"
 else
-CLEANUP="true"
+CLEANUP ?= "true"
```
This allows me to run `make kind CLEANUP=false` locally.

wallrj commented on Makefile:
```diff
@@ -108,3 +109,6 @@ CONTROLLER_GEN=$(GOBIN)/controller-gen
 endif
+
+certmanager-install:
+	kubectl apply --validate=false -f https://github.com/jetstack/cert-manager/releases/download/v0.11.0/cert-manager.yaml
```
This makes it easier to install everything in an existing cluster.
It's dupliated in the e2e test setup, which is a shame.

wallrj commented on api/v1alpha1/validation.go:
```diff
@@ -7,6 +7,9 @@ import (
 )
 
+//+kubebuilder:webhook:path=/validate-etc-improbable-io-v1-etcdclusters,verbs=create;update,mutating=false,failurePolicy=fail,groups=etcd.improbable.io,resources=etcdclusters;etcdpeers,versions=v1,name=validation.etcd.improbable.io
+//+kubebuilder:webhook:path=/default-etc-improbable-io-v1-etcdclusters,verbs=create;update,mutating=true,failurePolicy=fail,groups=etcd.improbable.io,resources=etcdclusters;etcdpeers,versions=v1,name=default.etcd.improbable.io
```
These govern the generated webhook configurations below.

wallrj commented on config/certmanager/certificate.yaml:
```diff
@@ -1,14 +1,14 @@
 # The following manifests contain a self-signed issuer CR and a certificate CR.
 # More document can be found at https://docs.cert-manager.io
-apiVersion: certmanager.k8s.io/v1alpha1
+apiVersion: cert-manager.io/v1alpha2
```
Kubebuilder 0.3 installed certmanager resources which are incompatible with CM 0.11. See https://github.com/kubernetes-sigs/kubebuilder/issues/1093

wallrj commented on config/default/kustomization.yaml:
```diff
@@ -1,12 +1,12 @@
 # Adds namespace to all resources.
-namespace: etcd-cluster-operator-system
+namespace: eco-system
```
I shortened the namespace and the name prefix below because otherwise we end up with service DNS names (and hence certificate common names) which are longer than 64 characters....which cert-manager rejects.

```
Resource: "cert-manager.io/v1alpha2, Resource=certificates", GroupVersionKind: "cert-manager.io/v1alpha2, Kind=Certificate"
Name: "etcd-cluster-operator-serving-cert", Namespace: "etcd-cluster-operator-system"
Object: &{map["apiVersion":"cert-manager.io/v1alpha2" "kind":"Certificate" "metadata":map["annotations":map["kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"cert-manager.io/v1alpha2\",\"kind\":\"Certificate\",\"metadata\":{\"annotations\":{},\"name\":\"etcd-cluster-operator-serving-cert\",\"namespace\":\"etcd-cluster-operator-system\"},\"spec\":{\"commonName\":\"$(SERVICE_NAME).$(SERVICE_NAMESPACE).svc\",\"dnsNames\":[\"$(SERVICE_NAME).$(SERVICE_NAMESPACE).svc.cluster.local\"],\"issuerRef\":{\"kind\":\"Issuer\",\"name\":\"selfsigned-issuer\"},\"secretName\":\"webhook-server-cert\"}}\n"] "creationTimestamp":"2019-10-30T12:32:42Z" "generation":'\x01' "name":"etcd-cluster-operator-serving-cert" "namespace":"etcd-cluster-operator-system" "resourceVersion":"15401" "selfLink":"/apis/cert-manager.io/v1alpha2/namespaces/etcd-cluster-operator-system/certificates/etcd-cluster-operator-serving-cert" "uid":"20621af3-a988-4540-875e-1ca8b4b704ab"] "spec":map["commonName":"$(SERVICE_NAME).$(SERVICE_NAMESPACE).svc" "dnsNames":["$(SERVICE_NAME).$(SERVICE_NAMESPACE).svc.cluster.local"] "issuerRef":map["kind":"Issuer" "name":"selfsigned-issuer"] "secretName":"webhook-server-cert"] "status":map["conditions":[map["lastTransitionTime":"2019-10-30T12:32:42Z" "message":"Waiting for CertificateRequest \"etcd-cluster-operator-serving-cert-1917100924\" to complete" "reason":"InProgress" "status":"False" "type":"Ready"]]]]}
for: "STDIN": admission webhook "webhook.cert-manager.io" denied the request: spec.commonName: Too long: must have at most 64 characters

```

adamhosier replied:
`ecosystem` nice 😁

wallrj commented on config/default/manager_auth_proxy_patch.yaml:
```diff
@@ -3,8 +3,8 @@
-  name: manager
-  namespace: default
+  name: controller-manager
+  namespace: system
```
The name and namespace had been altered in https://github.com/improbable-eng/etcd-cluster-operator/pull/27/files#diff-4693cc41eb3660fc1d3af8a11304f44c 
(I can't remember why)

But that breaks all the other kustomize patches that come with kubebuilder, 
so I've reverted that change.

Also updated the e2e tests accordingly.

wallrj commented on config/manager/manager.yaml:
```diff
@@ -1,8 +1,8 @@
-  name: manager
-  namespace: default
+  name: controller-manager
+  namespace: system
```
See above.

wallrj commented on config/webhook/manifests.yaml:
```diff
@@ -0,0 +1,54 @@
+    - UPDATE
+    resources:
+    - etcdclusters
+    - etcdpeers
```
Auto generated.

wallrj commented on config/webhook/manifests.yaml:
```diff
@@ -0,0 +1,54 @@
+  name: validating-webhook-configuration
+webhooks:
+- clientConfig:
+    caBundle: Cg==
```
This gets filled in by cert-manager ca injector: https://docs.cert-manager.io/en/latest/reference/cainjector.html

adamhosier commented on README.md:
```diff
@@ -40,6 +40,21 @@ make deploy
 
 This will leave changes on disk in your `config` directory, take care not to commit them.
 
+### Deploying webhooks / Installing Cert-Manager
```
i'm not sure from reading this whether the webhooks are required or not. can a user chose to not have them if they don't want the functionality? are there any k8s version restrictions

adamhosier commented on README.md:
```diff
@@ -40,6 +40,21 @@ make deploy
+
+These HTTP webhook requests, require SSL certificates to be configured for each of webhook request type,
+and the easiest way to set this up, is to install [Cert-Manager](https://book.kubebuilder.io/cronjob-tutorial/running-webhook.html#cert-manager),
+before you deploy the Operator.
```
maybe specify that these are self-signed certs and will not require any external dependency on any CA (assuming that's right)?

adamhosier commented on main.go:
```diff
@@ -62,6 +64,21 @@ func main() {
 		os.Exit(1)
 	}
 	// +kubebuilder:scaffold:builder
+	if os.Getenv("ENABLE_WEBHOOKS") != "false" {
```
seems like a strangely named var to me, i'd expect that if I did not set `ENABLE_WEBHOOKS` then the webhooks would not be enabled - however i don't think thats currently the case. would it make more sense to call this en var `DISABLE_WEBHOOKS` with the condition `os.Getenv("DISABLE_WEBOOKS") != ""`

wallrj replied:
I copied that from the Kuberbuilder book: https://book.kubebuilder.io/cronjob-tutorial/main-revisited.html

But I agree that it is confusing so I'll change it.

wallrj commented:
@adamhosier I think I've addressed your code review comments:
 * Added more detailed documentation about the webhook requirement and about how cert-manager creates the SSL certificates.
 * Change `ENABLE_WEBHOOKS` to `DISABLE_WEBHOOKS`.
 * Fixed some function documentation.

Please take another look.

wallrj commented on internal/test/e2e/testdata/validation/etcdpeer_missing_storageclassname.yaml:
```diff
@@ -0,0 +1,13 @@
+# A peer with a missing storageclassname
```
Done.
